File: sourcegen-generator-java/src/test/java/io/micronaut/sourcegen/javapoet/write/GenericInferenceRegressionTest.java
/*
 * Copyright 2017-2026 original authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.micronaut.sourcegen.javapoet.write;

import io.micronaut.sourcegen.JavaPoetSourceGenerator;
import io.micronaut.sourcegen.model.ClassDef;
import io.micronaut.sourcegen.model.ClassTypeDef;
import io.micronaut.sourcegen.model.MethodDef;
import io.micronaut.sourcegen.model.ObjectDef;
import io.micronaut.sourcegen.model.TypeDef;
import org.junit.jupiter.api.Test;

import javax.lang.model.element.Modifier;
import java.io.IOException;
import java.io.StringWriter;

import static io.micronaut.sourcegen.javapoet.write.JavaCompileAssertions.assertCompiles;
import static org.junit.jupiter.api.Assertions.assertEquals;

/**
 * Regression coverage for bounds resolved in the caller and callee scopes.
 */
class GenericInferenceRegressionTest {

    @Test
    void inferredBoundKeepsTheReceiversFixedVariable() throws IOException {
        var x = TypeDef.variable("X", TypeDef.of(Number.class));
        var u = TypeDef.variable("U", x);
        var identity = MethodDef.builder("identity").addTypeVariable(u).addParameter("value", u).returns(u)
            .build((self, p) -> p.getFirst().returning());
        var target = ClassDef.builder("test.FixedTarget").addTypeVariable(x).addMethod(identity).build();
        var v = TypeDef.variable("V", TypeDef.of(Number.class));
        var caller = ClassDef.builder("test.FixedCaller").addTypeVariable(v)
            .addMethod(MethodDef.builder("call").addParameter("target", TypeDef.parameterized(target.asTypeDef(), v))
                .addParameter("value", Number.class).returns(Number.class)
                .build((self, p) -> p.getFirst().invoke(identity, p.get(1)).returning())).build();
        assertCompiles(
            assertSource(target, """
                package test;

                import java.lang.Number;

                class FixedTarget<X extends Number> {
                  <U extends X> U identity(U value) {
                    return value;
                  }
                }
                """),
            assertSource(caller, """
                package test;

                import java.lang.Number;

                class FixedCaller<V extends Number> {
                  Number call(FixedTarget<V> target, Number value) {
                    return (Number) target.identity((V) value);
                  }
                }
                """)
        );
    }

    @Test
    void boxedArgumentConvertsToTheReceiversFixedVariable() throws IOException {
        var x = TypeDef.variable("X", TypeDef.of(Number.class));
        var u = TypeDef.variable("U", x);
        var identity = MethodDef.builder("identity").addTypeVariable(u).addParameter("value", u).returns(u)
            .build((self, p) -> p.getFirst().returning());
        var target = ClassDef.builder("test.BoxedFixedTarget").addTypeVariable(x).addMethod(identity).build();
        var v = TypeDef.variable("V", TypeDef.of(Number.class));
        var caller = ClassDef.builder("test.BoxedFixedCaller").addTypeVariable(v)
            .addMethod(MethodDef.builder("call").addParameter("target", TypeDef.parameterized(target.asTypeDef(), v))
                .addParameter("value", int.class).returns(Number.class)
                .build((self, p) -> p.getFirst().invoke(identity, p.get(1)).returning())).build();
        assertCompiles(
            assertSource(target, """
                package test;

                import java.lang.Number;

                class BoxedFixedTarget<X extends Number> {
                  <U extends X> U identity(U value) {
                    return value;
                  }
                }
                """),
            assertSource(caller, """
                package test;

                import java.lang.Integer;
                import java.lang.Number;

                class BoxedFixedCaller<V extends Number> {
                  Number call(BoxedFixedTarget<V> target, int value) {
                    return (Number) target.identity((V) (Integer) value);
                  }
                }
                """)
        );
    }

    @Test
    void chainedRecursiveBoundDoesNotRetainCalleeVariables() throws IOException {
        var t = TypeDef.variable("T", TypeDef.parameterized(ClassTypeDef.of(Comparable.class), TypeDef.variable("T")));
        var u = TypeDef.variable("U", t);
        var identity = MethodDef.builder("identity").addTypeVariable(t).addTypeVariable(u).addParameter("value", u).returns(u)
            .build((self, p) -> p.getFirst().returning());
        var def = ClassDef.builder("test.ChainedRecursive")
            .addMethod(identity)
            .addMethod(MethodDef.builder("call").addParameter("value", Object.class).returns(Object.class)
                .build((self, p) -> self.invoke(identity, p.getFirst()).returning())).build();
        assertCompiles(assertSource(def, """
            package test;

            import java.lang.Comparable;
            import java.lang.Object;

            class ChainedRecursive {
              <T extends Comparable<T>, U extends T> U identity(U value) {
                return value;
              }

              Object call(Object value) {
                return this.identity((Comparable) value);
              }
            }
            """));
    }

    private static String assertSource(ObjectDef objectDef, String expected) throws IOException {
        try (StringWriter writer = new StringWriter()) {
            new JavaPoetSourceGenerator().write(objectDef, writer);
            String source = writer.toString();
            assertEquals(expected, source);
            return source;
        }
    }
}

File: sourcegen-generator-kotlin/src/test/kotlin/io/micronaut/sourcegen/GenericInferenceRegressionTest.kt
/*
 * Copyright 2017-2026 original authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.micronaut.sourcegen

import io.micronaut.sourcegen.KotlinCompileAssertions.assertSource
import io.micronaut.sourcegen.model.ClassDef
import io.micronaut.sourcegen.model.ClassTypeDef
import io.micronaut.sourcegen.model.MethodDef
import io.micronaut.sourcegen.model.TypeDef
import io.micronaut.sourcegen.KotlinCompileAssertions.assertCompiles
import org.junit.jupiter.api.Test
import java.util.concurrent.ForkJoinPool
import java.util.concurrent.ForkJoinTask
import javax.lang.model.element.Modifier

/**
 * Regression coverage for bound conversion at generic invocation sites.
 */
class GenericInferenceRegressionTest {

    @Test
    fun chainedCalleeBoundIsResolvedBeforeCasting() {
        val t = TypeDef.variable("T", TypeDef.of(CharSequence::class.java))
        val u = TypeDef.variable("U", t)
        val identity = MethodDef.builder("chained").addModifiers(Modifier.PUBLIC, Modifier.STATIC)
            .addTypeVariable(t).addTypeVariable(u).addParameter("value", u).returns(u).build()
        val def = ClassDef.builder("test.ChainedCall")
            .addMethod(MethodDef.builder("call").addParameter("value", TypeDef.OBJECT).returns(TypeDef.OBJECT)
                .build { _, p -> ClassTypeDef.of(Calls::class.java).invokeStatic(identity, p[0]).returning() }).build()
        assertCompiles(assertSource(def, """
            |package test
            |
            |import io.micronaut.sourcegen.GenericInferenceRegressionTest
            |import kotlin.Any
            |import kotlin.CharSequence
            |
            |public open class ChainedCall {
            |  public open fun call(`value`: Any): Any {
            |    return GenericInferenceRegressionTest.Calls.chained(`value` as CharSequence)
            |  }
            |}
            |""".trimMargin()))
    }

    @Test
    fun recursiveCalleeBoundKeepsItsRawBound() {
        val t = TypeDef.variable("T", TypeDef.parameterized(ClassTypeDef.of(Comparable::class.java), TypeDef.variable("T")))
        val identity = MethodDef.builder("recursive").addModifiers(Modifier.PUBLIC, Modifier.STATIC)
            .addTypeVariable(t).addParameter("value", t).returns(t).build()
        val def = ClassDef.builder("test.RecursiveCall")
            .addMethod(MethodDef.builder("call").addParameter("value", TypeDef.OBJECT).returns(TypeDef.OBJECT)
                .build { _, p -> ClassTypeDef.of(Calls::class.java).invokeStatic(identity, p[0]).returning() }).build()
        assertCompiles(assertSource(def, """
            |package test
            |
            |import io.micronaut.sourcegen.GenericInferenceRegressionTest
            |import kotlin.Any
            |import kotlin.Comparable
            |
            |public open class RecursiveCall {
            |  public open fun call(`value`: Any): Any {
            |    return GenericInferenceRegressionTest.Calls.recursive(`value` as Comparable<Any>)
            |  }
            |}
            |""".trimMargin()))
    }

    @Test
    fun inferredVariableSatisfiesEveryIntersectionBound() {
        val t = TypeDef.variable("T", TypeDef.of(CharSequence::class.java), TypeDef.of(Runnable::class.java))
        val identity = MethodDef.builder("intersection").addModifiers(Modifier.PUBLIC, Modifier.STATIC)
            .addTypeVariable(t).addParameter("value", t).returns(t).build()
        val def = ClassDef.builder("test.IntersectionCall")
            .addMethod(MethodDef.builder("call").addParameter("value", TypeDef.OBJECT).returns(TypeDef.OBJECT)
                .build { _, p -> ClassTypeDef.of(Calls::class.java).invokeStatic(identity, p[0]).returning() }).build()
        assertCompiles(assertSource(def, """
            |package test
            |
            |import io.micronaut.sourcegen.GenericInferenceRegressionTest
            |import java.lang.Runnable
            |import kotlin.Any
            |import kotlin.CharSequence
            |
            |public open class IntersectionCall {
            |  public open fun call(`value`: Any): Any {
            |    return GenericInferenceRegressionTest.Calls.intersection(run {
            |      `value` as CharSequence
            |      `value` as Runnable
            |      `value`
            |    })
            |  }
            |}
            |""".trimMargin()))
    }

    @Test
    fun inferredBoundCastsTheWholeConditionalArgument() {
        val task = TypeDef.parameterized(ClassTypeDef.of(ForkJoinTask::class.java), TypeDef.wildcard())
        val t = TypeDef.variable("T", task)
        val submit = MethodDef.builder("submit").addModifiers(Modifier.PUBLIC).addTypeVariable(t)
            .addParameter("task", t).returns(t).build()
        val def = ClassDef.builder("test.ConditionalBoundCall")
            .addMethod(MethodDef.builder("call").addParameter("pool", ForkJoinPool::class.java)
                .addParameter("flag", TypeDef.Primitive.BOOLEAN).addParameter("left", TypeDef.OBJECT).addParameter("right", TypeDef.OBJECT)
                .returns(TypeDef.OBJECT)
                .build { _, p -> p[0].invoke(submit, p[1].isTrue().doIfElse(p[2], p[3])).returning() }).build()
        assertCompiles(assertSource(def, """
            |package test
            |
            |import java.util.concurrent.ForkJoinPool
            |import java.util.concurrent.ForkJoinTask
            |import kotlin.Any
            |import kotlin.Boolean
            |
            |public open class ConditionalBoundCall {
            |  public open fun call(
            |    pool: ForkJoinPool,
            |    flag: Boolean,
            |    left: Any,
            |    right: Any,
            |  ): Any {
            |    return pool.submit((if (flag) left else right) as ForkJoinTask<*>)
            |  }
            |}
            |""".trimMargin()))
    }

    @Test
    fun inferredBoundUsesReceiverArguments() {
        val x = TypeDef.variable("X", TypeDef.of(CharSequence::class.java))
        val u = TypeDef.variable("U", x)
        val identity = MethodDef.builder("identity").addModifiers(Modifier.PUBLIC).addTypeVariable(u)
            .addParameter("value", u).returns(u).build()
        val receiver = TypeDef.parameterized(ClassTypeDef.of(Receiver::class.java), TypeDef.STRING)
        val def = ClassDef.builder("test.ReceiverBoundCall")
            .addMethod(MethodDef.builder("call").addParameter("target", receiver).addParameter("value", TypeDef.OBJECT)
                .returns(TypeDef.OBJECT)
                .build { _, p -> p[0].invoke(identity, p[1]).returning() }).build()
        assertCompiles(assertSource(def, """
            |package test
            |
            |import io.micronaut.sourcegen.GenericInferenceRegressionTest
            |import kotlin.Any
            |import kotlin.String
            |
            |public open class ReceiverBoundCall {
            |  public open fun call(target: GenericInferenceRegressionTest.Receiver<String>, `value`: Any): Any {
            |    return target.identity(`value` as String)
            |  }
            |}
            |""".trimMargin()))
    }

    @Test
    fun recursiveBoundInsideACollectionIsPreserved() {
        val t = TypeDef.variable("T", TypeDef.parameterized(ClassTypeDef.of(Comparable::class.java), TypeDef.variable("T")))
        val first = MethodDef.builder("recursiveList").addModifiers(Modifier.PUBLIC, Modifier.STATIC).addTypeVariable(t)
            .addParameter("values", TypeDef.parameterized(ClassTypeDef.of(List::class.java), t)).returns(t).build()
        val def = ClassDef.builder("test.RecursiveCollectionCall")
            .addMethod(MethodDef.builder("call").addParameter("value", TypeDef.OBJECT).returns(TypeDef.OBJECT)
                .build { _, p -> ClassTypeDef.of(Calls::class.java).invokeStatic(first, p[0]).returning() }).build()
        assertCompiles(assertSource(def, """
            |package test
            |
            |import io.micronaut.sourcegen.GenericInferenceRegressionTest
            |import kotlin.Any
            |import kotlin.Comparable
            |import kotlin.collections.List
            |
            |public open class RecursiveCollectionCall {
            |  public open fun call(`value`: Any): Any {
            |    return GenericInferenceRegressionTest.Calls.recursiveList(`value` as List<Comparable<Any>>)
            |  }
            |}
            |""".trimMargin()))
    }

    class Calls {
        companion object {
            @JvmStatic
            fun <T : CharSequence, U : T> chained(value: U): U = value

            @JvmStatic
            fun <T : Comparable<T>> recursive(value: T): T = value

            @JvmStatic
            fun <T : Comparable<T>> recursiveList(values: List<T>): T = values.first()

            @JvmStatic
            fun <T> intersection(value: T): T where T : CharSequence, T : Runnable = value
        }
    }

    class Receiver<X : CharSequence> {
        fun <U : X> identity(value: U): U = value
    }
}
